### PR TITLE
tridiagonal eigensolver: uniformize usage of end index

### LINKED ISSUE #708: Consistently exclude the end index from ranges
The following algorithms:

- general sub-matrix-matrix multiplication
- column/row sub-matrix permutations
- tridiagonal eigensolver
- given rotations application

use ranges where the end index is included in public and internal APIs.

### PR DESCRIPTION
**Pull request by rasolca**
Closes #708

Refactored tridiagonal eigensolver to use open ranges.

- [x] #854
- [x] remove some duplications (Waiting for #766 before moving some functions)
- [x] clang-format (only at the end to simplify diffs)

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/index_manipulation.h:37]:
```suggestion
void initIndex(const SizeType i_begin, const SizeType i_end, Matrix<SizeType, D>& index) {
```

File: include/dlaf/permutations/general.h
Comment (msimberg) [inline include/dlaf/permutations/general.h:92]: Presumably this TODO means you want to wait for #854 to be merged before this PR?

File: include/dlaf/eigensolver/tridiag_solver/merge.h
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:281]: This is peak bikeshedding, and I'm happy to take this discussion to a less formal coffee discussion, but shouldn't this stay `constexpr auto backend`? This is a local variable and we've been naming variables with snake case. `B` in `template <Backend B>` is a template parameter and we've been naming them with capital case.

**Review by msimberg — CHANGES_REQUESTED** [review]
My only requested change is removing `inline` from the function template since it's not necessary, the rest can be ignored (especially the `backend` naming comment).

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h
Comment (albestro) [inline include/dlaf/eigensolver/tridiag_solver/index_manipulation.h:31]: Do you think it might be worth an `DLAF_ASSERT_HEAVY` here for checking that `i_end - 1` does not fall outside the valid range for `dist`?

**Review by albestro — APPROVED** [review]

File: include/dlaf/permutations/general.h
Comment (albestro) [reply include/dlaf/permutations/general.h:92]: We will synchronise with @rasolca, but an idea might be that I rebase #854 and make it a fork of this one (so we merge that one in this one).

File: include/dlaf/eigensolver/tridiag_solver/merge.h
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:281]: you are right... This is the result of a find replace that went off of its scope... 😉

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/index_manipulation.h:31]: tileSize should already assert in some way... or am I wrong?

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/index_manipulation.h:37]: 👍🏻 I simply moved the function from a file to another and I didn't notice it...

File: include/dlaf/permutations/general.h
Comment (rasolca) [reply include/dlaf/permutations/general.h:92]: My idea was to merge this PR and when ready #854. But I'm fine with @albestro suggestion as well.

File: include/dlaf/eigensolver/tridiag_solver/index_manipulation.h
Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/index_manipulation.h]: removed in 2 places ( 410eb191249072191829666465607711ab784abf )

**Comment by rasolca** [comment]
cscs-ci run

File: include/dlaf/eigensolver/tridiag_solver/rot.h
Comment (RMeli) [inline include/dlaf/eigensolver/tridiag_solver/rot.h:177]: Add `SingleThreadedBlasScope` in this lambda.

Comment (rasolca) [reply include/dlaf/eigensolver/tridiag_solver/rot.h:180]: Done during the merge https://github.com/eth-cscs/DLA-Future/pull/843/commits/7332a5fe1dc6ae1c99f995fea84521391f740f05

**Comment by rasolca** [comment]
cscs-ci run

**Review by msimberg — APPROVED** [review]

**Pull request merged**
